xtra-libp2p/src/endpoint_monitor.rs

```diff
@@ -0,0 +1,268 @@
+                    dropped: connections_dropped.clone(),
+                })
+                .await
+            {
+                tracing::error!(
+                    "Unable to reach a subscriber that is supposed to be supervised: {e:#}"
+                );
+            }
+        }
+    }
+
+    async fn notify_listen_addresses_added(&mut self, addresses_added: HashSet<Multiaddr>) {
+        if addresses_added.is_empty() {
+            return;
+        }
+
+        for subscriber in &self.subscribers.listen_address_added {
+            if let Err(e) = subscriber
+                .send(ListenAddressAdded {
+                    added: addresses_added.clone(),
+                })
+                .await
+            {
+                tracing::error!(
+                    "Unable to reach a subscriber that is supposed to be supervised: {e:#}"
+                );
+            }
+        }
+    }
+
+    async fn notify_listen_addresses_removed(&mut self, addresses_removed: HashSet<Multiaddr>) {
+        if addresses_removed.is_empty() {
+            return;
+        }
+
+        for subscriber in &self.subscribers.listen_address_removed {
+            if let Err(e) = subscriber
+                .send(ListenAddressRemoved {
+                    removed: addresses_removed.clone(),
+                })
+                .await
+            {
+                tracing::error!(
+                    "Unable to reach a subscriber that is supposed to be supervised: {e:#}"
+                );
+            }
+        }
+    }
+}
+
+#[async_trait]
+impl xtra::Actor for Actor {
+    type Stop = Error;
+
+    async fn started(&mut self, ctx: &mut xtra::Context<Self>) {
+        let this = ctx.address().expect("we are alive");
+
+        self.tasks
+            .add(this.send_interval(CHECK_INTERVAL, || CheckConnection));
+    }
+
+    async fn stopped(self) -> Self::Stop {
+        self.stop_reason.unwrap_or(Error::Unspecified)
+    }
+}
+
+#[xtra_productivity]
+impl Actor {
+    async fn handle(&mut self, msg: Error, ctx: &mut xtra::Context<Self>) {
+        self.stop_reason = Some(msg);
+        ctx.stop();
+    }
+
+    async fn handle(&mut self, _msg: CheckConnection, ctx: &mut xtra::Context<Self>) {
+        if let Err(e) = self.monitor_connections().await {
+            self.stop_reason = Some(e);
+            ctx.stop();
+        }
+    }
+}
+
+#[derive(thiserror::Error, Copy, Clone, Debug)]
+pub enum Error {
+    #[error("Endpoint actor is disconnected")]
+    NoEndpoint,
+    #[error("Stop reason was not specified")]
+    Unspecified,
+}
+
+#[derive(Clone, Copy)]
+pub struct CheckConnection;
+
+pub struct ConnectionsEstablished {
+    pub established: HashSet<PeerId>,
+}
+
+impl xtra::Message for ConnectionsEstablished {
+    type Result = ();
+}
+
+pub struct ConnectionsDropped {
+    pub dropped: HashSet<PeerId>,
+}
+
+impl xtra::Message for ConnectionsDropped {
+    type Result = ();
+}
+
+pub struct ListenAddressAdded {
```

Comment:
```suggestion
pub struct ListenAddressesAdded {
```

it is multiple, same as for the connections